server/controllers/logstreaming_controller.go

```diff
@@ -117,17 +117,34 @@ func (j *LogStreamingController) GetLogStreamWS(w http.ResponseWriter, r *http.R
 	}
 
 	c, err := j.WebsocketHandler.Upgrade(w, r, nil)
-
 	if err != nil {
 		j.Logger.Warn("Failed to upgrade websocket: %s", err)
 		return
 	}
 
-	defer c.Close()
+	// Add a reader goroutine to listen for socket.close() events.
+	go func() {
+		for {
+			_, _, err := c.ReadMessage()
+			if err != nil {
+				j.Logger.Warn("Failed to read WS message: %s", err)
+				return
+			}
+		}
+	}()
 
 	pull := pullInfo.String()
 
-	err = j.ProjectCommandOutputHandler.Receive(pull, func(msg string) error {
+	// Buffer size = 10 allows for any dealys in ws connection message.
+	wsChannel := make(chan string, 10)
```

Comment:
I'm not sure why do we need buffer size here? What delay in connection impacts the buffer?

Reply:
When we get a new message in the ProjectCmdOutput channel, we iterate through all the channels in ReceiverBuffer and forward the message. However, if a unbuffered channel is not ready to receive msg, we delete the chan here: 
https://github.com/lyft/atlantis/blob/22f56bbee06b1c17c3021d917dd87495b5c2641c/server/handlers/project_command_output_handler.go#L120-L127
This is what happens when the UI gets stuck. When one of the message takes longer to send (callback function) through the ws and the receiving channel is not ready to receive the message (shown below), we just delete the channel which causes the UI to get stuck since there's no channel that's forwarding the messages to the UI. https://github.com/lyft/atlantis/blob/22f56bbee06b1c17c3021d917dd87495b5c2641c/server/handlers/project_command_output_handler.go#L69-L73
Making the channel buffered (size 10 for example) allows the writeLogLine() method to keep on pushing messages even if the channel is not ready to receive the messages yet. If it's still not ready after 10 enqued messaged, we could assume that something's wrong with the ch and delete the channel. We could increase the buffer size imo

Reply:
This was the fix to the UI bug btw 😅

Reply:
Hm, ok. If that's the case we can revert this value back to 1000 as it was in the original code. And we can revisit the implementation later. Receiving part of log streaming is a bit too complicated for what it is doing.

Reply:
True! I'll revisit the architecture in a later PR.